src/proxyvisor.coffee

```diff
@@ -0,0 +1,313 @@
+				status: 'Provisioned'
+			resinApi.post
+				resource: 'device'
+				body: d
+				customOptions:
+					apikey: apiKey
+			.then (dev) ->
+				deviceForDB = {
+					uuid: uuid
+					appId: d.application
+					device_type: d.device_type
+					deviceId: dev.id
+					name: dev.name
+					status: d.status
+					logs_channel: d.logs_channel
+				}
+				knex('dependentDevice').insert(deviceForDB)
+				.then ->
+					res.status(201).send(dev)
+	)
+	.catch (err) ->
+		console.error("Error on #{req.method} #{url.parse(req.url).pathname}", err, err.stack)
+		res.status(503).send(err?.message or err or 'Unknown error')
+
+router.get '/v1/devices/:uuid', (req, res) ->
+	uuid = req.params.uuid
+	knex('dependentDevice').select().where({ uuid })
+	.then ([ device ]) ->
+		return res.status(404).send('Device not found') if !device?
+		res.json(parseDeviceFields(device))
+	.catch (err) ->
+		console.error("Error on #{req.method} #{url.parse(req.url).pathname}", err, err.stack)
+		res.status(503).send(err?.message or err or 'Unknown error')
+
+router.post '/v1/devices/:uuid/logs', (req, res) ->
+	uuid = req.params.uuid
+	m = {
+		message: req.body.message
+		timestamp: req.body.timestamp or Date.now()
+	}
+	m.isSystem = req.body.isSystem if req.body.isSystem?
+
+	knex('dependentDevice').select().where({ uuid })
+	.then ([ device ]) ->
+		return res.status(404).send('Device not found') if !device?
+		pubnub.publish({ channel: "device-#{device.logs_channel}-logs", message: m })
+		res.status(202).send('OK')
+	.catch (err) ->
+		console.error("Error on #{req.method} #{url.parse(req.url).pathname}", err, err.stack)
+		res.status(503).send(err?.message or err or 'Unknown error')
+
+validStringOrUndefined = (s) ->
+	_.isUndefined(s) or !_.isEmpty(s)
+validObjectOrUndefined = (o) ->
+	_.isUndefined(o) or _.isObject(o)
+
+router.put '/v1/devices/:uuid', (req, res) ->
+	uuid = req.params.uuid
+	{ status, is_online, commit, environment, config } = req.body
+	if !_.isUndefined(is_online) and !_.isBoolean(is_online)
+		res.status(400).send('is_online must be a boolean')
+		return
+	if !validStringOrUndefined(status)
+		res.status(400).send('status must be a non-empty string')
+		return
+	if !validStringOrUndefined(commit)
+		res.status(400).send('commit must be a non-empty string')
+		return
+	if !validObjectOrUndefined(environment)
+		res.status(400).send('environment must be an object')
+		return
+	if !validObjectOrUndefined(config)
+		res.status(400).send('config must be an object')
+		return
+	environment = JSON.stringify(environment)
+	config = JSON.stringify(config)
+
+	Promise.join(
+		utils.getConfig('apiKey')
+		knex('dependentDevice').select().where({ uuid })
+		(apiKey, [ device ]) ->
+			throw new Error('apikey not found') if !apiKey?
+			return res.status(404).send('Device not found') if !device?
+			resinApi.patch
+				resource: 'device'
+				id: device.deviceId
+				body: _.pick({ status, is_online, commit }, (v) -> !_.isUndefined(v))
```

Comment:
This can also be `_.negate(_.isUndefined))`, you could also extract it as `isNotUndefined = _.negate(_.isUndefined))` (or `isDefined`)